### PR TITLE
Switch from lizard to lz4 and update compressors comparison.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Switch from lizard to lz4 and update compressors comparison.

From lzbench, it looks like we may see a ~15% improvement in compression and decompression speed _and_ produce slightly smaller files.

Actual impact on Stripe codebase:

* 2% reduction in cache size.
* ~13% improvement in _file and AST_ compression speed.
* Small (~3%?) improvement in GlobalState compression speed.
* Negligible improvement in GlobalState decompression speed.
* Negligible change on indexing on a warm cache.

### Motivation

Speed++.

### Test plan

Covered by existing tests.

### CODE REVIEW

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Iw2svOY7xUx16Q
→ https://go/builds/bui_Iw2sL61wcVx1sR

File: docs/compressors.md [hunk 1]
```diff
@@ -1,63 +1,56 @@
...
-| lzrw 15-Jul-1991 -4     |    9606 us |    6748 us |     1470050 | 40.38 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| tornado 0.6a -2         |   11920 us |    6932 us |     1251726 | 34.38 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| quicklz 1.5.0 -2        |   12054 us |    7179 us |     1187468 | 32.62 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| density 0.14.2 -3       |    9822 us |   10258 us |     1606968 | 44.14 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| tornado 0.6a -3         |   19214 us |   11493 us |     1022778 | 28.09 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-
-Note: LZ5 has a downsize though. It's really bad in compressing small pieces of data. <1KB data sizes normally _grow_ after compression.
+| lz4fast 1.9.3 -17       |    8368 us |    2023 us |     3555129 | 57.23 | uncompressed-gs-rle|
+| lzsse8fast 2019-04-18   |   30112 us |    2065 us |     2301226 | 37.05 | uncompressed-gs-rle|
+| lzsse4fast 2019-04-18   |   29646 us |    2189 us |     2306934 | 37.14 | uncompressed-gs-rle|
```
Comment (froydnj) [inline docs/compressors.md:14]: I can't really make a connection between the cited settings above and the line in this table which corresponds to those settings.  Is it one of these, or a different line?

File: docs/compressors.md [hunk 2]
```diff
@@ -1,63 +1,56 @@
...
-| tornado 0.6a -1         |   10191 us |    6431 us |     1463980 | 40.21 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| lzrw 15-Jul-1991 -4     |    9606 us |    6748 us |     1470050 | 40.38 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| tornado 0.6a -2         |   11920 us |    6932 us |     1251726 | 34.38 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| quicklz 1.5.0 -2        |   12054 us |    7179 us |     1187468 | 32.62 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| density 0.14.2 -3       |    9822 us |   10258 us |     1606968 | 44.14 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-| tornado 0.6a -3         |   19214 us |   11493 us |     1022778 | 28.09 | sorbet-stdlib-state-a2f6fab208ee500774c6faa9c62c228c3b21fe08|
-
-Note: LZ5 has a downsize though. It's really bad in compressing small pieces of data. <1KB data sizes normally _grow_ after compression.
+| lz4fast 1.9.3 -17       |    8368 us |    2023 us |     3555129 | 57.23 | uncompressed-gs-rle|
```
Comment (froydnj) [inline docs/compressors.md:12]: Is it possible to include the hash of the checkout in the filename, so people can try to replicate these results?

File: core/serialize/serialize.cc [hunk 3]
```diff
@@ -73,15 +72,15 @@ vector<u1> Pickler::result(int compressionDegree) {
...
-    const size_t maxDstSize = Lizard_compressBound(data.size());
+    const size_t maxDstSize = LZ4_compressBound(data.size());
     vector<u1> compressedData;
     compressedData.resize(2048 + maxDstSize); // give extra room for compression
                                               // Lizard_compressBound returns size of data if compression
                                               // succeeds. It seems to be written for big inputs
                                               // and returns too small sizes for small inputs,
                                               // where compressed size is bigger than original size
-    int resultCode = Lizard_compress((const char *)data.data(), (char *)(compressedData.data() + SIZE_BYTES * 2),
-                                     data.size(), (compressedData.size() - SIZE_BYTES * 2), compressionDegree);
```
Comment (froydnj) [inline core/serialize/serialize.cc:84]: I guess we no longer use `compressionDegree`?  Is it possible to change that, so we are explicitly selecting the compression settings LZ4 will use?

**Review by froydnj — APPROVED** [review]
I have some questions, but nothing blocking.

File: docs/compressors.md (hunk 1 above)
Comment (jvilk-stripe) [reply docs/compressors.md:14]: Good question. I'm using `lz4 1.9.3` as we _do_ care a bit about compression speed _and_ size of the compressed output. These variants either have slow compression or bad compression ratio. I can point out / bold the line we are using.

File: docs/compressors.md (hunk 2 above)
Comment (jvilk-stripe) [reply docs/compressors.md:12]: Sure, although I had to modify Sorbet to actually output an uncompressed global state.

File: core/serialize/serialize.cc (hunk 3 above)
Comment (jvilk-stripe) [reply core/serialize/serialize.cc:84]: Yeah, we're using the default. I can make it explicit?

Comment (froydnj) [reply core/serialize/serialize.cc:84]: I am a fan of explicit so some future upgrade doesn't silently change the defaults out from under us.  Please change.

**Pull request merged**